x-pack/filebeat/input/httpjson/request.go

```diff
@@ -146,38 +167,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	}
+	return *newUrl, nil
+
+}
+
+func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher) error {
+	var (
+		n                 int
+		ids               []string
+		err               error
+		split             *split
+		urlCopy           url.URL
+		urlString         string
+		httpResp          *http.Response
+		intermediateResps []*http.Response
+		finalResps        []*http.Response
+	)
+	rfSize := len(r.requestFactories)
+	for i, rf := range r.requestFactories {
+		// iterate over collected ids from last response
+		if i == 0 {
+			// perform and store regular call responses
+			httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+			if err != nil {
+				return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
+			}
+			if rfSize == 1 {
+				finalResps = append(finalResps, httpResp)
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, finalResps, true)
+				if err != nil {
+					return err
+				}
+				continue
+			} else {
+				intermediateResps = append(intermediateResps, httpResp)
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+				if err != nil {
+					return err
+				}
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, intermediateResps, false)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			if len(ids) == 0 {
+				n = 0
+				continue
+			}
+			urlCopy = rf.url
+			urlString = rf.url.String()
+			// perform request over collected ids
+			for _, id := range ids {
+				// reformat urls of requestFactory using ids
+				rf.url, err = generateNewUrl(rf.replace, urlString, id)
+				if err != nil {
+					return fmt.Errorf("failed to generate new URL: %w", err)
+				}
+
+				// collect data from new urls
+				httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+				if err != nil {
+					return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
+				}
+				// store data according to response type
+				if i == rfSize-1 && len(ids) != 0 {
+					finalResps = append(finalResps, httpResp)
+				} else {
+					intermediateResps = append(intermediateResps, httpResp)
+				}
+			}
+			rf.url = urlCopy
+
+			if i == rfSize-1 {
+				split = r.responseProcessor.split
+				r.responseProcessor.split = rf.split
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, finalResps, true)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+				continue
+			} else {
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+				if err != nil {
+					return err
+				}
+				split = r.responseProcessor.split
+				r.responseProcessor.split = rf.split
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, intermediateResps, false)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+			}
+		}
 	}
+
 	defer httpResp.Body.Close()
```

Comment:
This doesn't need to be deferred since it's a final act anyway.

Reply:
Right